docs/rfc/rfc-016-node-architecture.md

```diff
@@ -0,0 +1,78 @@
+- Should draw a distinction between a service and a process.  
+  - The blocksync (and statesync) service, i.e. supplying information for those trying to catch up should only start running once a node has caught up i.e. after running the blocksync and/or state sync *processes*
+  - The blocksync and state sync processes have defined termination clauses that inform the orchestrator when they are done and where they finished.
+    - One way of achieving this would be that every process both passes and returns the `State` object
+  - Consensus doesn't necessarily terminate and thus the line between service and process is more blurry
+    - However, one could view the services as running the mempool, evidence and peer state routines which provide information for the consensus engine to run.
+    - And just as blocksync uses peer heights to decide when it is finished, Consensus could use peer heights to decide that it has fallen behind and needs to terminate for blocksync to start. 
+  - This distinction needs to be communicated in the p2p layer. For example, if a node is state syncing it shouldn't receive requests for snapshots.
+- Should know when to switch from one state advancing mechanism to another. The most challenging being to know when to fall back to blocksync.
+  - Either the orchestrator instructs consensus to stop if it falls *n* blocks behind consensus (*n* being infinity if, for example, the blocksync service is switched off and we never want consensus to stop)
+
+The orchestrator allows for some deal of variablity in how a node is constructed. Does it just run blocksync, shadowing the head of the chain and be highly available for querying. Does it rely on state sync at all? An important question that arises from this dynamicism is we ideally want to encourage nodes to provide as much of their resources as possible so that their is a healthy amount of providers to consumers. Do we make all services compulsory or allow for them to be disabled? Arguably it's possible that a user forks the codebase and rips out the blocksync code because they want to reduce bandwidth so this is more a question of how easy do we want to make this for users.
+
+### Block Executor
+
+The block executor is an important component that is currently used by both consensus and blocksync to perform the execution grunt work with the application: applying blocks and updating state. Principally, I think it should be the only component that can write (and possibly even read) the block and state stores and we should clean up other references to the storage engine if we can. This would mean:
+
+- The reactors Consensus, BlockSync and StateSync should all import the executor for advancing state ie.  `ApplyBlock` and `BootstrapState` 
+- Pruning should also be a concern of the block executor as well as `FinalizeBlock` and `Commit`. This can simplify consensus to focus just on the consensus part.
```

Comment:
I strongly agree with this section. I think that this `Executor`, which I think we've occasionally referred to as the `ChainManager` would be a much nicer abstraction. This would allow a single point of synchronization for advancing the chain. This would allow all users of this type to attempt to produce the next height independently of any other block producers.  

I think a reasonable interface for this type would be the following:

```go
type ChainManager interface {
  // LatestState always returns the latest state known by the chain manager.
  LatestState() (state.State, error)
  // UpdateState directly sets the state of the chain known by the ChainManager.
  UpdateState(state.State) error
 // ApplyBlock accepts a block and the commit that justifies that block. If the height of the chain is far beyond the height of the block being applied, ApplyBlock performs no work and returns an error indicating to the caller that the chain has advanced beyond the applied block. ApplyBlock validates that the block is correct using its latest copy of the state.
  ApplyBlock(types.Block, types.Commit) error
}
```

Consensus and BlockSync both attempt to fetch blocks/seenCommits the controller feeds them into `ApplyBlock`. Whether or not to run BlockSync or Consensus at instant can become a decision between how far behind the `ChainManager`'s view of the chain is in comparison to the node's peers. 

This would obviously have implications on how we architect the consensus and blocksync packages as both would need to be able to switch to running with any supplied `State`. 

The 'controller' code could be a simple loop that gets the next block from either blocksync or consensus and puts it into `ApplyBlock`.

Reply:
> The 'controller' code could be a simple loop that gets the next block from either blocksync or consensus and puts it into `ApplyBlock`.

I quite like this formulation, most especially for the latter reason. At the risk of getting into the weeds, one concern the interface needs to address is that this controller needs to be able to detect "edges", i.e., moments when the latest state has changed. Otherwise, the loop "spins".

But for it _not_ to spin, we have to have some sense of time: The controller needs not only to know what the latest state is, but when it has "changed" (I use quotes because what matters is not that the value necessarily change, but that someone poked at it in an actionable way).

That detail aside, however, I think this is the right approach. (And that is solvable—it's an important but entirely tractable API design question)

Reply:
Yes I like the phrasing of a single point of synchronization. I would modify your API slightly and say that ApplyBlock should also return the new state that derives from the application. 
```
ApplyBlock(types.Block, types.Commit) (state.State, error)
```

Reply:
I think `ApplyBlock` which executes the block against the application will always result in State being updated.

Reply:
So, in my envisioning of this, different block-applying mechanisms could be run in parallel. They would all try to call ApplyBlock with the latest block. One may beat the other to apply a block and result in no change in State. I think both APIs could suit both of our proposed ways of implementing this.

Reply:
In which case you'd still want to return the state because it would be more advanced in height than the one the mechanism initially had. I'm not sure that I like the idea of these mechanisms running in parallel. I'd prefer orchestration where 1 of the 3 are used